cpp/include/cudf/io/text/detail/tile_state.hpp

```diff
@@ -0,0 +1,174 @@
+  }
+
+  __device__ inline void set_inclusive_prefix(cudf::size_type tile_idx, T value)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+    cub::ThreadStore<cub::STORE_CG>(tile_inclusive + offset, value);
+    __threadfence();
+    cub::ThreadStore<cub::STORE_CG>(tile_status + offset, scan_tile_status::inclusive);
+  }
+
+  __device__ inline T get_prefix(cudf::size_type tile_idx, scan_tile_status& status)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+
+    while ((status = cub::ThreadLoad<cub::LOAD_CG>(tile_status + offset)) ==
+           scan_tile_status::invalid) {
+      __threadfence();
+    }
+
+    if (status == scan_tile_status::partial) {
+      return cub::ThreadLoad<cub::LOAD_CG>(tile_partial + offset);
+    } else {
+      return cub::ThreadLoad<cub::LOAD_CG>(tile_inclusive + offset);
+    }
+  }
+
+  __device__ inline T get_inclusive_prefix(cudf::size_type tile_idx)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+    while (cub::ThreadLoad<cub::LOAD_CG>(tile_status + offset) != scan_tile_status::inclusive) {
+      __threadfence();
+    }
+    return cub::ThreadLoad<cub::LOAD_CG>(tile_inclusive + offset);
+  }
+};
+
+template <typename T>
+struct scan_tile_state {
+  rmm::device_uvector<scan_tile_status> tile_status;
+  rmm::device_uvector<T> tile_state_partial;
+  rmm::device_uvector<T> tile_state_inclusive;
+
+  scan_tile_state(cudf::size_type num_tiles,
+                  rmm::cuda_stream_view stream,
+                  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+    : tile_status(rmm::device_uvector<scan_tile_status>(num_tiles, stream, mr)),
+      tile_state_partial(rmm::device_uvector<T>(num_tiles, stream, mr)),
+      tile_state_inclusive(rmm::device_uvector<T>(num_tiles, stream, mr))
+  {
+  }
+
+  operator scan_tile_state_view<T>()
+  {
+    return scan_tile_state_view<T>{tile_status.size(),
+                                   tile_status.data(),
+                                   tile_state_partial.data(),
+                                   tile_state_inclusive.data()};
+  }
+
+  inline void set_seed_async(T const seed, rmm::cuda_stream_view stream)
+  {
+    auto x = tile_status.size();
+    auto y = scan_tile_status::inclusive;
+    tile_state_inclusive.set_element_async(x - 1, seed, stream);
+    tile_status.set_element_async(x - 1, y, stream);
+  }
+
+  // T back_element(rmm::cuda_stream_view stream) const { return tile_state.back_element(stream); }
+
+  inline T get_inclusive_prefix(cudf::size_type tile_idx, rmm::cuda_stream_view stream) const
+  {
+    auto const offset = (tile_idx + tile_status.size()) % tile_status.size();
+    return tile_state_inclusive.element(offset, stream);
+  }
+};
+
+template <typename T>
+struct scan_tile_state_callback {
+  using WarpReduce = cub::WarpReduce<T>;
+
+  struct _TempStorage {
+    typename WarpReduce::TempStorage reduce;
```

Comment:
Are we using the `WarpReduce` anywhere?

Reply:
leftover from alternative implementation of the partial-prefix summation, which was slower. I'll remove it.

Reply:
Interesting! So, [this part here](https://github.com/rapidsai/cudf/pull/8702#discussion_r682532135) turned out to be slower when implemented using a WarpReduce instead of having a single thread iterate over the items?

Reply:
Yep.